### DISCUSSION TITLE
JackTrip w/ OSC

### DISCUSSION THREAD

**Discussion by cchafe** [jacktrip/jacktrip#815] [Ideas]
There is interest in adding a side-channel for [OSC](https://ccrma.stanford.edu/groups/osc/index.html) to JackTrip. It's something we could toss around here and I can plan an interested parties meeting. There have been one-off mods in the past to add EEG and CV signals.

**Comment by OConnorStP** [comment]
I'm using OSC a lot these days...

Would it be for self-hosted/classic use or just Studio?  I can see it being useful for the Studio controls, but I'm drawing a blank as to how this would work in Classic, since that's more a protocol and doesn't offer much in the way of in-flight controls. 

Which brings up the topic of forking Classic and Studio versions of Jacktrip?  Jacktrip the protocol (Classic) vs Jacktrip the product (Studio)?  

Something that might be cool is an OSC tunnel that rides the Classic Jacktrip stream.  With that a person could transport OSC to/from any port on either end without having to punch holes in their firewalls.  That could lead to a whole life for Jacktrip that doesn't have anything to do with audio -- bidirectional high performance OSC.

**Comment by OConnorStP** [comment]
I thought of another cool protocol to throw into the tunnel.  Ableton Link:

> [Ableton Link](https://ableton.com/link) is a technology that synchronizes musical beat, tempo, phase, and start/stop commands across multiple applications running on one or more devices. Applications on **devices connected to a local network** discover each other automatically and form a musical session in which each participant can perform independently: anyone can start or stop while still staying in time. Anyone can change the tempo, the others will follow. Anyone can join or leave without disrupting the session.

I've wanted to learn of somebody bypassing the phrase "devices connected to a local network" ever since it came out, since all the music I do is over the wire.  Here's the key notion -- Link sends multicast UDP messages to 224.76.78.75:20808 on the local area network to communicate with other peers.   Is it possible to bridge those packets up to UDP and drop them down in other networks via Jacktrip??  *That* would be cool.

Don't read on if that's impossible...  :-)
________________________

Link is an open-source project, with a nice pile of documentation.  Here are some links to Link:

Top of the pile:  https://www.ableton.com/en/link/

Enabled products:  https://www.ableton.com/en/link/products/  (Max supports link -- think of the fun to be had...  Maybe Miller could add it to PD?)

Top-level description on GitHub:  https://ableton.github.io/link/  

GitHub repository:  https://github.com/Ableton/link

**Comment by syneme** [comment]
H, I've always used oscgroups alongside jacktrip. we built that into artsmesh. we use it for triggering score animation and remote processing, but also for chat (chat over osc). However, I tend to think about what should ideally be an integrated function and what should be parallel? For example, we have an artsmesh user server which keeps track of user status, IP/port info, etc, but then the oscgroups is another 'group' which parallels that. I wish we could integrate those. Oscgroups has a NAT whole punching algo which I always wondered how we could use in jacktrip at the same time. Then p2p would be so much smoother in JT. Ken

**Reply by syneme** [reply]
port mapping is one step beyond which 99 percent of musicians will not go. k.
